File: app/src/main/java/vassar/cmpu203/vassardiningapp/view/MenuSelectFragment.java
package vassar.cmpu203.vassardiningapp.view;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.view.MenuProvider;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Lifecycle;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.snackbar.Snackbar;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import vassar.cmpu203.vassardiningapp.R;
import vassar.cmpu203.vassardiningapp.databinding.FragmentMenuSelectBinding;
import vassar.cmpu203.vassardiningapp.model.MealtimeMenu;

public class MenuSelectFragment extends Fragment implements IMenuSelectView, MenuProvider {

    private final IMenuSelectView.Listener listener;
    private FragmentMenuSelectBinding binding;
    private ExpandableMealtimeAdapter itemsAdapter;
    private String cafe;
    private LocalDate date = LocalDate.now();

    public MenuSelectFragment(IMenuSelectView.Listener listener) {
        this.listener = listener;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup parent, @Nullable Bundle savedInstanceState) {
        binding = FragmentMenuSelectBinding.inflate(inflater);
        requireActivity().addMenuProvider(this, getViewLifecycleOwner(), Lifecycle.State.RESUMED);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getActivity().setTitle("Menu");

        LinearLayoutManager layoutManager = new LinearLayoutManager(view.getContext());
        layoutManager.setOrientation(RecyclerView.VERTICAL);
        binding.mealtimeRecycler.setLayoutManager(layoutManager);
        itemsAdapter = new ExpandableMealtimeAdapter(new ArrayList<>(), getContext(), listener, this);
        binding.mealtimeRecycler.setAdapter(itemsAdapter);

        populateSpinner(view, binding.cafeSpinner, R.array.cafes);

        binding.textDate.setText(date.toString());
        refreshMenu();
    }

    /**
     * Translates the given cafe name to fit the HTTP request.
     *
     * @param cafe The name of the cafe to be translated.
     * @return The translated name for the given cafe.
     */
    private String translateCafe(String cafe) {
        String translatedCafe = cafe;
        switch (cafe) {
            case "Deece":
                translatedCafe = "gordon";
                break;
            case "Retreat":
                translatedCafe = "the-retreat";
                break;
            case "Street Eats":
                translatedCafe = "food-truck";
                break;
        }
        return translatedCafe;
    }

    /**
     * Populates a Spinner view with a given array of strings.
     *
     * @param view           The view containing the Spinner.
     * @param spinner        The Spinner to be populated.
     * @param textArrayResId The resource ID of the array of strings to be used as the Spinner's options.
     */
    private void populateSpinner(View view, Spinner spinner, int textArrayResId) {
        ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(view.getContext(), textArrayResId, android.R.layout.simple_spinner_item);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(adapter);

        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                cafe = translateCafe(binding.cafeSpinner.getSelectedItem().toString());
                listener.loadData(cafe, date, MenuSelectFragment.this);
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }

    /**
     * Updates the menus displayed in the fragment.
     *
     * @param menus Menus to be displayed.
     */
    @Override
    public void updateMenuItems(List<MealtimeMenu> menus) {
        if (menus.isEmpty()) {
            Snackbar.make(binding.getRoot(), "Menu not found", Snackbar.LENGTH_SHORT).show();
        }
        itemsAdapter.setMenus(menus);
        refreshMenu();
    }

    /**
     * Refreshes the menu displayed in the fragment.
     */
    @SuppressLint("NotifyDataSetChanged")
    @Override
    public void refreshMenu() {
        itemsAdapter.notifyDataSetChanged();
    }

    @Override
    public void onCreateMenu(@NonNull Menu menu, @NonNull MenuInflater menuInflater) {
        menuInflater.inflate(R.menu.main_action_bar, menu);

        // Ensure filter buttons are displayed properly
        MenuItem favoriteFilterButton = menu.findItem(R.id.favorite_filter_button);
        favoriteFilterButton.setIcon(listener.getUser().isFavoriteFiltered()
                ? R.drawable.ic_white_filled_heart : R.drawable.ic_white_empty_heart);
        MenuItem applyFilterButton = menu.findItem(R.id.apply_filter_button);
        applyFilterButton.setIcon(listener.getUser().isRestrictionFiltered()
                ? R.drawable.ic_filled_dining : R.drawable.ic_empty_dining);
    }

    @Override
    public boolean onMenuItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.previous_day_button) {
            date = date.minusDays(1);
            binding.textDate.setText(date.toString());
            listener.loadData(cafe, date, this);
        } else if (id == R.id.next_day_button) {
            date = date.plusDays(1);
            binding.textDate.setText(date.toString());
            listener.loadData(cafe, date, this);
        } else if (id == R.id.favorite_filter_button) {
            listener.getUser().toggleFavoriteFilter();
            item.setIcon(listener.getUser().isFavoriteFiltered()
                    ? R.drawable.ic_white_filled_heart : R.drawable.ic_white_empty_heart);
            listener.updateVisibleMenu(this);
            return true;
        } else if (id == R.id.apply_filter_button) {
            listener.getUser().toggleRestrictionFilter();
            item.setIcon(listener.getUser().isRestrictionFiltered()
                    ? R.drawable.ic_filled_dining : R.drawable.ic_empty_dining);
            listener.updateVisibleMenu(this);
        }
        return false;
    }
}

File: androidApp/src/main/java/vassar/cmpu203/vassardiningapp/view/MenuSelectFragment.kt
package vassar.cmpu203.vassardiningapp.view

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.*
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import vassar.cmpu203.vassardiningapp.R
import vassar.cmpu203.vassardiningapp.databinding.FragmentMenuSelectBinding
import vassar.cmpu203.vassardiningapp.model.MealtimeMenu
import java.time.LocalDate

private const val TITLE = "Menu"

class MenuSelectFragment(private val listener: IMenuSelectView.Listener) :
        Fragment(), IMenuSelectView, MenuProvider {

    private lateinit var binding: FragmentMenuSelectBinding
    private lateinit var itemsAdapter: ExpandableMealtimeAdapter
    private lateinit var cafe: String
    private var date: LocalDate = LocalDate.now()

    override fun onCreateView(inflater: LayoutInflater, parent: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = FragmentMenuSelectBinding.inflate(inflater)
        requireActivity().addMenuProvider(this, viewLifecycleOwner, Lifecycle.State.RESUMED)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        activity?.title = TITLE

        val layoutManager = LinearLayoutManager(view.context).apply {
            orientation = RecyclerView.VERTICAL
        }
        binding.mealtimeRecycler.layoutManager = layoutManager
        itemsAdapter = ExpandableMealtimeAdapter(listOf(), requireContext(), listener, this)
        binding.mealtimeRecycler.adapter = itemsAdapter

        populateSpinner(view, binding.cafeSpinner, R.array.cafes)

        binding.textDate.text = date.toString()
        refreshMenu()
    }

    private fun translateCafe(cafe: String) = when (cafe) {
        "Deece" -> "gordon"
        "Retreat" -> "the-retreat"
        "Street Eats" -> "food-truck"
        else -> cafe
    }

    private fun populateSpinner(view: View, spinner: Spinner, textArrayResId: Int) {
        ArrayAdapter.createFromResource(
                view.context,
                textArrayResId,
                android.R.layout.simple_spinner_item
        ).also { adapter ->
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            spinner.adapter = adapter
        }

        spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                cafe = translateCafe(binding.cafeSpinner.selectedItem.toString())
                listener.loadData(cafe, date, this@MenuSelectFragment)
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    override fun updateMenuItems(menus: List<MealtimeMenu>) {
        if (menus.isEmpty()) {
            Snackbar.make(binding.root, "Menu not found", Snackbar.LENGTH_SHORT).show()
        }
        itemsAdapter.setMenus(menus)
        refreshMenu()
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun refreshMenu() {
        itemsAdapter.notifyDataSetChanged()
    }

    override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
        menuInflater.inflate(R.menu.main_action_bar, menu)

        // Ensure filter buttons are displayed properly
        val favoriteFilterButton = menu.findItem(R.id.favorite_filter_button)
        favoriteFilterButton.setIcon(
                if (listener.user.isFavoriteFiltered) R.drawable.ic_white_filled_heart else R.drawable.ic_white_empty_heart
        )
        val applyFilterButton = menu.findItem(R.id.apply_filter_button)
        applyFilterButton.setIcon(
                if (listener.user.isRestrictionFiltered) R.drawable.ic_filled_dining else R.drawable.ic_empty_dining
        )
    }

    override fun onMenuItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.previous_day_button -> {
                date = date.minusDays(1)
                binding.textDate.text = date.toString()
                listener.loadData(cafe, date, this)
                false
            }
            R.id.next_day_button -> {
                date = date.plusDays(1)
                binding.textDate.text = date.toString()
                listener.loadData(cafe, date, this)
                false
            }
            R.id.favorite_filter_button -> {
                listener.user.toggleFavoriteFilter()
                item.setIcon(
                        if (listener.user.isFavoriteFiltered) R.drawable.ic_white_filled_heart else R.drawable.ic_white_empty_heart
                )
                listener.updateVisibleMenu(this)
                true
            }
            R.id.apply_filter_button -> {
                listener.user.toggleRestrictionFilter()
                item.setIcon(
                        if (listener.user.isRestrictionFiltered) R.drawable.ic_filled_dining else R.drawable.ic_empty_dining
                )
                listener.updateVisibleMenu(this)
                false
            }
            else -> false
        }
    }
}
